app/v2/models/response/retrieveBalanceAndTransactions/FinanceDetails.scala

```diff
@@ -0,0 +1,117 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2.models.response.retrieveBalanceAndTransactions
+
+import play.api.libs.functional.syntax.toFunctionalBuilderOps
+import play.api.libs.json.{JsPath, Json, OWrites, Reads}
+import v2.models.response.retrieveBalanceAndTransactions.financeDetailTypes.mapping
+
+case class FinanceDetails(taxYear: String,
+                          documentId: String,
+                          chargeType: Option[String],
+                          mainType: Option[String],
+                          taxPeriodFrom: Option[String],
+                          taxPeriodTo: Option[String],
+                          contractAccountCategory: Option[String],
+                          contractAccount: Option[String],
+                          documentNumber: Option[String],
+                          documentNumberItem: Option[String],
+                          chargeReference: Option[String],
+                          mainTransaction: Option[String],
+                          subTransaction: Option[String],
+                          originalAmount: Option[BigDecimal],
+                          outstandingAmount: Option[BigDecimal],
+                          clearedAmount: Option[BigDecimal],
+                          accruedInterest: Option[BigDecimal],
+                          items: Seq[FinancialDetailsItem])
+
+object FinanceDetails {
+
+  implicit val reads: Reads[FinanceDetails] = (
+    (JsPath \ "taxYear").read[String] and
+      (JsPath \ "documentId").read[String] and
+      (JsPath \ "chargeType").readNullable[String] and
+      (JsPath \ "mainType").readNullable[String].map(c => getMainType(c)) and
+      (JsPath \ "taxPeriodFrom").readNullable[String] and
+      (JsPath \ "taxPeriodTo").readNullable[String] and
+      (JsPath \ "contractAccountCategory").readNullable[String] and
+      (JsPath \ "contractAccount").readNullable[String] and
+      (JsPath \ "sapDocumentNumber").readNullable[String] and
+      (JsPath \ "sapDocumentNumberItem").readNullable[String] and
+      (JsPath \ "chargeReference").readNullable[String] and
+      (JsPath \ "mainTransaction").readNullable[String] and
+      (JsPath \ "subTransaction").readNullable[String] and
+      (JsPath \ "originalAmount").readNullable[BigDecimal] and
+      (JsPath \ "outstandingAmount").readNullable[BigDecimal] and
+      (JsPath \ "clearedAmount").readNullable[BigDecimal] and
+      (JsPath \ "accruedInterest").readNullable[BigDecimal] and
+      (JsPath \ "items").read[Seq[FinancialDetailsItem]]
+  )(FinanceDetails.apply _)
+
+  private def getMainType(downstreamValue: Option[String]): Option[String] = {
+    mapping.get(downstreamValue.get) match {
+      case None    => None
+      case Some(s) => Option(s)
+      case _       => None
```

Comment:
Best avoid calling `get` on an `Option` that might be empty. Instead maybe something like:
```suggestion    
    downstreamValue.flatMap(x => mapping.get(x))
```